test/EtomicSwap.js

```diff
@@ -99,6 +116,95 @@ contract('EtomicSwap', function(accounts) {
         await this.swap.erc20Payment(...params).should.be.rejectedWith(EVMThrow);
     });
 
+    it('should allow to send ERC721 payment', async function () {
+        const lockTime = await currentEvmTime() + 1000;
+        const tokenId = 1; // Assuming token ID 1 is minted to accounts[0] in Erc721Token contract
+        const params = [
+            id,
+            accounts[1], // Receiver address
+            this.erc721token.address, // ERC721 token contract address
+            tokenId, // Token ID
+            secretHash,
+            lockTime
+        ];
+
+        // approve the swap contract to transfer the token on behalf of accounts[0]
+        await this.erc721token.approve(this.swap.address, tokenId, { from: accounts[0] });
```

Comment:
Hmm I see.

We cant avoid calling approve if we want to keep token on Etomic Swap address on some process stage.
I wouldn't like to over complicate `onERC721Received` logic as it is a part of safeTransferFrom execution. I mean additional complex logic can potentially lead to reentrancy risks, while safeTransferFrom was designed to mitigate some such risks.
In addition complex computations in onERC721Received can increase the gas cost.

What do you think about calling additional event related to Nft in `erc721Payment` like `AwaitingToken` and encoding id with paymentHash in data.

In `onERC721Received` we can check `nfts[id].state` and `nfts[id].paymentHash` if requirements are met `then return this.onERC721Received.selector`.

I pushed rough impl example in separate [branch](https://github.com/KomodoPlatform/etomic-swap/commit/9b8b078ddc3ca5ac6a38ba1d20e504ebe1301968)
Such checks will revert direct `safeTransferFrom` and keep Etomic Swap contracts behavior simple and predictable.

Reply:
> In addition complex computations in onERC721Received can increase the gas cost.

We already have overhead in gas costs for users because they need to make approve call before calling `erc721Payment`. Handling HTLC creation logic in `onERC721Received` will simply move the existing code from one fn to another, likely keeping gas costs for the user the same (or even lower because no `approve` is required).

> I mean additional complex logic can potentially lead to reentrancy risks

We are not going to call another smart contract in `onERC721Received`, just change internal `EtomicSwap` state. How is it supposed to lead to reentrancy risks?

> In onERC721Received we can check nfts[id].state and nfts[id].paymentHash if requirements are met then return this.onERC721Received.selector.

This actually adds more computations during payment creation, incurring additional gas costs (should be negligible, but still :slightly_smiling_face: )

Also, you can see actual gas usage in tests, e.g:
```js
        let tx_info = await this.swap.erc1155Payment(...params, { from: accounts[0] }).should.be.fulfilled;
        console.log(tx_info);
```

The object has `receipt` inner object with `gasUsed` field:
```
gasUsed: 95826,
```
So, you can compare `gasUsed` when different approaches are used. If it turns out that `onERC721Received` approach really increases `gasUsed`, I will reconsider my proposal.

Reply:
@artemii235 yeah. you are right, calling safeTransferFrom directly is more gas effective [gist](https://gist.github.com/laruh/8bf0a5cc903925a272e5e1fe2258f3de). Will add changes for erc721 and erc1155.

Reply:
Done